### ISSUE TITLE
Using shared memory in Numba and Cupy functions 

### ISSUE THREAD

**Issue by Mitko88** [numba/numba#5754]
##  Feature request

I tried to execute a Cupy function with shared memory arrays and cuda.jit, but it seems it is not possible at the moment. Currently, I need to write numpy/cupy functions by myself if I want to use shared memory arrays.

**Comment by esc** [comment]
@Mitko88 thanks for submitting this, I have labelled it as a feature request.

**Comment by gmarkall** [comment]
Indeed, it is not possible at present to call CuPy functions from inside `@cuda.jit` functions at present. How this should work is not entirely obvious to me at present - some design work and thinking would be needed to determine what is possible and how it should be implemented.

**Comment by leofang** [comment]
xref: https://github.com/cupy/cupy/issues/3380

**Comment by sgbaird** [comment]
@Mitko88 or anyone else, do you know of a source for functions from numpy to use with cuda.jit and shared memory? I too have been writing them myself and feel like I must be repeating work that others have already done.

**Comment by gmarkall** [comment]
@sgbaird I think people have variously been implementing them as they need them, but there's no coherent collection of them. One of the issues is that there are implementations of many functions from NumPy in Numba, but they only work for the CPU target due to some architectural limitations in Numba. I'm hoping that over the next release or two we can resolve these issues and make it possible to call most NumPy functions inside CUDA kernels.

**Comment by sgbaird** [comment]
@gmarkall it definitely seems like a non-trivial task while ensuring that performance is retained. I struggle a lot with replacing NumPy functions that either do local array allocation (e.g. `np.zeroes()` or `np.concatenate()`) or return an array that's a different size than the inputs (`np.diff()`). I've been trying to use cuda.local.array for the former, but haven't had much success. I imagine for the latter I just need to initialize it higher up the stack trace and pass it in as an argument. Any thoughts? While NumPy (and CuPy) compatibility is in progress, I think a guide to rewriting NumPy functions for `cuda.jit` could be really helpful in the meantime. I'd be happy to do it, but I'm stuck on a few of these things. If that guide then had a link to somewhere people could post functions and short examples, I think that could be really useful in the interim.

**Comment by sgbaird** [comment]
Ooh, maybe of some use! https://the-algorithms.com/
I've also found a lot of source material at: https://geeksforgeeks.com/
Seems like educational platforms have quite the treasure trove of slow, vanilla Python code that can be accelerated via Numba and Numba/CUDA.
